src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,50 @@
+package duke;
+
+import java.util.Scanner;
+
+import duke.exceptions.DukeException;
+
+
+public class Duke {
+    private Ui ui;
+    private Parser parser;
+    private boolean isEnded;
+
+    public Duke() {
+        this.ui = new Ui();
+        this.parser = new Parser(ui);
+        this.isEnded = false;
+    }
+
+    private void run() {
+        String command;
+        Scanner sc = new Scanner(System.in);
+
+        ui.greet();
+        while (!this.isEnded) {
+            command = sc.nextLine();
+            try {
+                String response = parser.handler(command);
+                if (response.equals("Bye. Hope to see you again soon!\n")) {
+                    isEnded = true;
+                }
+                System.out.println(response);
+            } catch (DukeException e) {
+                System.out.println(ui.printException(e));
```

Comment:
Seems a little quirky that the method named `printException` has to be wrapped inside a `println` - perhaps the UI can handle the printing directly?